### ISSUE TITLE
Add optional expected exception types to failureResultOf

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6380]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#6380|
|Type|enhancement|
|Created|2013-03-18 20:57:10Z|
|Branch|https://github.com/twisted/twisted/tree/failureResultOf-optional-types-6380|

It would be nice to be able to pass optional errors to twisted.trial.unittest.TestCase.failureResultOf to check which errors are supposed to have been raised similar to (like assertRaises)

Split from [#6377](https://github.com/twisted/twisted/issues/6377)

<details><summary>Searchable metadata</summary>

```
trac-id__6380 6380
type__enhancement enhancement
reporter__cyli cyli
priority__normal normal
milestone__ 
branch__branches_failureResultOf_optional_types_6380 branches/failureResultOf-optional-types-6380
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1363640230000000 1363640230000000
changetime__1364370327000000 1364370327000000
version__None None
owner__cyli cyli

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli commented|
|-|-|

(In [37627]) Add optional expected failures to failureResultOf.  refs [#6380](https://github.com/twisted/twisted/issues/6380)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

(In [37629]) Fix bug - Failure.check([]) also returns None. refs [#6380](https://github.com/twisted/twisted/issues/6380)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli removed owner|
|-|-|

[buildbot results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/failureResultOf-optional-types-6380)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince set owner to @cyli|
|-|-|

1.There is one [twistedchecker error](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/479/steps/run-twistedchecker/logs/new%20twistedchecker%20errors).
 2. The names of `test_withFailureResult` and `test_withExpectedFailureResult` seem reversed.
 3. It would be good to test that the that the initial message is what is expected, in the case where no type, one type and multiple types are passed. 
 4. I guess that it is more complicate than it is worth, to list things as `"ex1, ex2 or ex3"`. That might be worth a follow up ticket, to implement a utility function somewhere in `twisted.python`.

Please merge after addressing 1+2+3.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

(In [37830]) Address review comments with regards to variable and method names and extra tests on the error message.  refs [#6380](https://github.com/twisted/twisted/issues/6380)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [37832]) Fix tests for python 3.  refs [#6380](https://github.com/twisted/twisted/issues/6380)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

1+2+3 have been addressed in r37830

Filed [#6393](https://github.com/twisted/twisted/issues/6393) to address (4)

[buildbot results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/failureResultOf-optional-types-6380) look good.

Going to merge.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli set status to `closed`|
|-|-|

(In [37834]) Merge failureResultOf-optional-types-6380: expected types of failureResultOf

Author: cyli
Reviewer: tom.prince
Fixes: [#6380](https://github.com/twisted/twisted/issues/6380)

Allow failureResultOf to take optional expected failure types, so that if an unexpected failure occurs, the failureResultOf assertion will fail.
